### PR TITLE
nixos/parsedmarc: Fix build and runtime

### PR DESCRIPTION
**Pull request by e1mo**
## Description of changes

Currently the `parsedmarc` Module is not working. This is due to:
1. `smtp.to` having defaults causing evaluation errors.
2. The default `settings.elasticsearch.hosts` (if elasticsearch is to be provisioned) not being a valid URL, causing errors.
3. `python311Packages.parsedmarc` not building successfully due to missing dependencies / incompatible versions of some dependencies.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/python-modules/parsedmarc/default.nix [hunk 1]
```diff
@@ -51,7 +51,8 @@ buildPythonPackage rec {
 
   postPatch = ''
     substituteInPlace pyproject.toml \
-      --replace "elasticsearch<7.14.0" "elasticsearch"
+      --replace "elasticsearch<7.14.0" "elasticsearch" \
+      --replace "elasticsearch-dsl==7.4.0" "elasticsearch-dsl"
   '';
```
Comment (mweinelt) [inline pkgs/development/python-modules/parsedmarc/default.nix:56]: Use `pythonRelaxDepsHook` instead.

```nix
  nativeBuildInputs = [
    pythonRelaxDepsHook
  ];

  pythonRelaxDeps = [
    "elasticseearch"
    "elasticsearch-dsl"
  ];

**Review by mweinelt — DISMISSED** [review]
Please update the attribute for the python-module changes to be prefixed with `python311Packages`, since `pythonPackages` is a reference for `python2.pkgs`.

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by mweinelt** [comment]
~1400 rebuilds on each -linux, ~600 on each -darwin. Likely a result from touching the elasticsearch module.

But going through a staging cycle to get the propagation fixed seems unreasonable.

File: pkgs/development/python-modules/elasticsearch/default.nix [hunk 2]
```diff
@@ -21,13 +21,10 @@ buildPythonPackage rec {
     sha256 = "sha256-nghBO+r/Oka8EMbFcGmoRwTfaqqTCFxzffB/WKKBG3g=";
   };
 
-  nativeBuildInputs = [
-    elastic-transport
-  ];
-
   propagatedBuildInputs = [
     urllib3
     certifi
```
Comment (mweinelt) [inline pkgs/development/python-modules/elasticsearch/default.nix:26]:
```suggestion
```

https://github.com/elastic/elasticsearch-py/blob/v8.9.0/setup.py#L54

**Comment by mweinelt** [comment]
Are the package changes also required for release-23.11? If not, could we only backport the module fixes, since they cause no rebuild?

**Comment by e1mo** [comment]
The backporting of the package fix is needed indeed :/

Parsedmarc tests on `release-23.11` do also fail with `ModuleNotFoundError: No module named 'elastic_transport'`.

**Comment by Janik-Haag** [comment]
more or less a duplicate of #279068

**Comment by e1mo** [comment]
I did bring this PR up to the current state of the module & package.
The latest update introduced `opensearch-py` new dependency which was not accounted for (but now is). The `msgraph-core` version is now pinned to `0.2.2`, thus parsedmarc is building again.

Since I would kinda like to have a working parsedmarc package and module again, I've kept the `smtp.to` patches that are a duplicate of #279068, as mentioned by @Janik-Haag.
Because those fixes on their own would not fix the broken build (and thus not fix broken deployments), I would prefer to go with this PR fixing all of breaking issues at once.

And maybe #285840 could get some fresh momentum to help get parsedmarc out of the persistent state of breakage?

**Comment by mweinelt** [comment]
Unfortunately overrides of libraries aren't possible within the python package set. That would only be possible if `parsedmarc` was built with `buildPythonApplication` outside `python3Packages`.

The reason is that libraries propagate their dependencies, and they ultimately all end up in a single PYTHONPATH that only supports exactly one version of a python module. With overrides in libraries it would be too easy to end up with multiple different propagated module versions, which would make the package set unmaintainable.

**Comment by e1mo** [comment]
Thanks for the explanation! I got kinda fooled by all tests passing (since we're not testing anything msgraph this makes sense tho) :/

Do you have any concerns regarding the changes introduced in #283337 (especially edf24111e45a1cc110fae47b9a8905b67e0983d7)? If not, I'd use them here too.

**Comment by mweinelt** [comment]
Yeah, that looks plausible.

**Review by SuperSandro2000 — APPROVED** [review]
LGTM

**Comment by e1mo** [comment]
I can't come up with anything that is still left open.
I also re-ran both parsedmarc test locally with the recent rebase and am happy to report that they're still passing :D

**Pull request merged**